### ISSUE TITLE
Model prediction results problem

### ISSUE THREAD

**Issue by Alande111** [ameraner/dsen2-cr#30]
I tested and found that the result appeared as an all-white image. The model I used is model_SARcarl.hdf5, and I only modified the batch size from 16 to 1. Afterward, I tried changing the threshold of use_cloud_mask and the value of scale, but neither of these changes affected the result. Can you help me identify where the issue might lie? Below is the log and results.

(SAR2) D:\pycharm\learn_pytorch\SAR-G\Code>python dsen2cr_main.py --predict D:\pycharm\learn_pytorch\SAR-G\Code\model\model_SARcarl.hdf5
Using TensorFlow backend.
2025-02-13 11:38:47.017036: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudart64_100.dll
WARNING:tensorflow:From dsen2cr_main.py:101: The name tf.ConfigProto is deprecated. Please use tf.compat.v1.ConfigProto instead.

WARNING:tensorflow:From dsen2cr_main.py:109: The name tf.Session is deprecated. Please use tf.compat.v1.Session instead.

2025-02-13 11:38:49.701461: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX AVX2
2025-02-13 11:38:49.704965: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library nvcuda.dll
2025-02-13 11:38:49.726480: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1618] Found device 0 with properties:
name: NVIDIA GeForce RTX 3060 Laptop GPU major: 8 minor: 6 memoryClockRate(GHz): 1.425
pciBusID: 0000:01:00.0
2025-02-13 11:38:49.726629: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudart64_100.dll
2025-02-13 11:38:49.732328: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cublas64_100.dll
2025-02-13 11:38:49.736236: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cufft64_100.dll
2025-02-13 11:38:49.738496: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library curand64_100.dll
2025-02-13 11:38:49.742623: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusolver64_100.dll
2025-02-13 11:38:49.745609: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusparse64_100.dll
2025-02-13 11:38:49.770056: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudnn64_7.dll
2025-02-13 11:38:49.770223: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1746] Adding visible gpu devices: 0
2025-02-13 11:39:53.966260: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1159] Device interconnect StreamExecutor with strength 1 edge matrix:
2025-02-13 11:39:53.966382: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1165]      0
2025-02-13 11:39:53.966546: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1178] 0:   N
2025-02-13 11:39:53.966936: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1304] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 4738 MB memory) -> phy
sical GPU (device: 0, name: NVIDIA GeForce RTX 3060 Laptop GPU, pci bus id: 0000:01:00.0, compute capability: 8.6)
WARNING:tensorflow:From dsen2cr_main.py:115: The name tf.set_random_seed is deprecated. Please use tf.compat.v1.set_random_seed instead.

WARNING:tensorflow:From D:\Anaconda3\envs\SAR2\lib\site-packages\keras\backend\tensorflow_backend.py:74: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_defaul
t_graph instead.

WARNING:tensorflow:From D:\Anaconda3\envs\SAR2\lib\site-packages\keras\backend\tensorflow_backend.py:517: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder ins
tead.

WARNING:tensorflow:From D:\Anaconda3\envs\SAR2\lib\site-packages\keras\backend\tensorflow_backend.py:4138: The name tf.random_uniform is deprecated. Please use tf.random.uniform instea
d.

WARNING:tensorflow:From D:\Anaconda3\envs\SAR2\lib\site-packages\keras\backend\tensorflow_backend.py:174: The name tf.get_default_session is deprecated. Please use tf.compat.v1.get_def
ault_session instead.

WARNING:tensorflow:From D:\Anaconda3\envs\SAR2\lib\site-packages\keras\backend\tensorflow_backend.py:190: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_var
iables instead.

WARNING:tensorflow:From D:\Anaconda3\envs\SAR2\lib\site-packages\keras\backend\tensorflow_backend.py:199: The name tf.is_variable_initialized is deprecated. Please use tf.compat.v1.is_
variable_initialized instead.

WARNING:tensorflow:From D:\Anaconda3\envs\SAR2\lib\site-packages\keras\backend\tensorflow_backend.py:206: The name tf.variables_initializer is deprecated. Please use tf.compat.v1.varia
bles_initializer instead.

WARNING:tensorflow:From D:\Anaconda3\envs\SAR2\lib\site-packages\keras\optimizers.py:790: The name tf.train.Optimizer is deprecated. Please use tf.compat.v1.train.Optimizer instead.

WARNING:tensorflow:From D:\pycharm\learn_pytorch\SAR-G\Code\tools\image_metrics.py:45: div (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Deprecated in favor of operator or tf.math.divide.
WARNING:tensorflow:From D:\Anaconda3\envs\SAR2\lib\site-packages\keras\backend\tensorflow_backend.py:1521: The name tf.log is deprecated. Please use tf.math.log instead.

Model compiled successfully!
Getting file lists
Number of train files found:  0
Number of validation files found:  0
Number of test files found:  2
Test file list: [['3', 's1', 's2_cloudFree', 's2_cloudy', 'ROIs2017_winter_61_p30.tif'], ['3', 's1', 's2_cloudFree', 's2_cloudy', 'ROIs2017_winter_8_p31.tif']]
Predicting using file: D:\pycharm\learn_pytorch\SAR-G\Code\model\model_SARcarl.hdf5
Using this model: test
WARNING: Folder {} exists and content may be overwritten!
WARNING: Folder {} exists and content may be overwritten!
Initializing generator for prediction and evaluation
Generator initialized
DataGenerator batch_size: 1
Storing evaluation metrics at  D:/pycharm/learn_pytorch/SAR-G/Code/output\testeval.csv
Processing file number  0
2025-02-13 11:39:55.499614: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudnn64_7.dll
2025-02-13 11:42:45.882060: W tensorflow/stream_executor/cuda/redzone_allocator.cc:312] Internal: Invoking ptxas not supported on Windows
Relying on driver to perform ptx compilation. This message will be only logged once.
2025-02-13 11:42:46.205622: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cublas64_100.dll
Generating quartet for  ROIs2017_winter_61_p30
D:\pycharm\learn_pytorch\SAR-G\Code\tools\dataIO.py:166: MatplotlibDeprecationWarning: Passing parameters norm and vmin/vmax simultaneously is deprecated since 3.3 and will become an e
rror two minor releases later. Please pass vmin/vmax directly to the norm when creating it.
  plt.imshow(image, cmap=cmap, norm=norm, vmin=-1, vmax=1)
Processing file number  1
Generating quartet for  ROIs2017_winter_8_p31
Prediction finished with success!

ROIs2017_winter_8_p31
![Image](https://github.com/user-attachments/assets/1db3b3bd-b2cf-463a-9a13-9ffe497e99df)

ROIs2017_winter_61_p30
![Image](https://github.com/user-attachments/assets/cbfaa6e3-e3df-4330-834e-7f30391eb674)

**Comment by ToddGoooo** [comment]
Hello, I got the same problem. Many of the pred images are totally white, and the min and max pixel values are same 255. Have you found the bug?

**Comment by ToddGoooo** [comment]
And when I use the trained model ,  I found a bug  in predict_dsen2cr.py file, you need to modify the json var params' string: batch_size to   var batch_size . The  original value is 1, and it can not work well.

**Comment by sllll12** [comment]
Hello, I would like to ask how your input file is organized? The file data I read is 0.

![Image](https://github.com/user-attachments/assets/e382f2f0-2fb5-44b0-8bf0-167016aee07d)

![Image](https://github.com/user-attachments/assets/23ff472c-2008-46ca-b532-f190ebbf04f1)

**Comment by Dargon2023** [comment]
这是我的其中一个预测结果，inputpred.png图片几乎是没有什么效果的，感觉是有什么问题。

This is one of my prediction results, inputpred.png image is almost no effect, what is the problem

![Image](https://github.com/user-attachments/assets/b84ddba2-2deb-4bfe-9535-b39538a49a90)

**Comment by ToddGoooo** [comment]
I have retrained this model, and the performance is better now. So I think something wrong in the auther's pretrained file.

**Comment by zhaojun-110** [comment]
Could you meet the loss is nan question? Or do you have changed some code? I am appreciated if you can reply. 

Using TensorFlow backend.
2025-05-01 19:34:58.851418: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
2025-05-01 19:34:58.911280: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:897] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2025-05-01 19:34:58.911338: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1405] Found device 0 with properties: 
name: NVIDIA GeForce RTX 4090 major: 8 minor: 9 memoryClockRate(GHz): 2.625
pciBusID: 0000:01:00.0
totalMemory: 23.64GiB freeMemory: 22.84GiB
2025-05-01 19:34:58.911348: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1484] Adding visible gpu devices: 0
2025-05-01 19:34:59.216550: I tensorflow/core/common_runtime/gpu/gpu_device.cc:965] Device interconnect StreamExecutor with strength 1 edge matrix:
2025-05-01 19:34:59.216567: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971]      0 
2025-05-01 19:34:59.216571: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 0:   N 
2025-05-01 19:34:59.216620: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 22145 MB memory) -> physical GPU (device: 0, name: NVIDIA GeForce RTX 4090, pci bus id: 0000:01:00.0, compute capability: 8.9)
Model compiled successfully!
Getting file lists
Number of train files found:  101615
Number of validation files found:  8623
Number of test files found:  7899
Training model name: DSen2-CR_001
WARNING: Folder {} exists and content may be overwritten!
Model information files created at  ../out/model_runs/DSen2-CR_001
Generator initialized
Generator initialized
Generator initialized
Training starts...
Epoch 1/8

   1/6350 [..............................] - ETA: 8:55:52 - loss: 2.6256 - carl_error: 2.6256 - cloud_mean_absolute_error: 1.3030 - cloud_mean_squared_error: 3.1282 - cloud_mean_sam: 0.9001 - cloud_mean_absolute_error_clear: 1.1958 - cloud_psnr: 9.0265 - cloud_root_mean_squared_error: 1.7687 - cloud_bandwise_root_mean_squared_error: 1.4042 - cloud_mean_absolute_error_covered: 1.4028 - cloud_ssim: 0.2091 - cloud_mean_sam_covered: 0.8848 - cloud_mean_sam_clear: 0.9424
   2/6350 [..............................] - ETA: 4:59:37 - loss: 2.3451 - carl_error: 2.3451 - cloud_mean_absolute_error: 1.1725 - cloud_mean_squared_error: 2.4982 - cloud_mean_sam: 0.7545 - cloud_mean_absolute_error_clear: 0.9490 - cloud_psnr: 10.1458 - cloud_root_mean_squared_error: 1.5677 - cloud_bandwise_root_mean_squared_error: 1.2737 - cloud_mean_absolute_error_covered: 1.3384 - cloud_ssim: 0.2914 - cloud_mean_sam_covered: 0.7378 - cloud_mean_sam_clear: 0.7488
   3/6350 [..............................] - ETA: 3:39:24 - loss: 1.8737 - carl_error: 1.8737 - cloud_mean_absolute_error: 0.9466 - cloud_mean_squared_error: 1.8716 - cloud_mean_sam: 0.5814 - cloud_mean_absolute_error_clear: 0.6806 - cloud_psnr: 12.1194 - cloud_root_mean_squared_error: 1.3073 - cloud_bandwise_root_mean_squared_error: 1.0420 - cloud_mean_absolute_error_covered: 1.1101 - cloud_ssim: 0.4261 - cloud_mean_sam_covered: 0.5967 - cloud_mean_sam_clear: 0.5159
   4/6350 [..............................] - ETA: 2:58:16 - loss: 3571587.9052 - carl_error: 3571587.9052 - cloud_mean_absolute_error: 1785793.9600 - cloud_mean_squared_error: 30703290941441.4023 - cloud_mean_sam: 0.6361 - cloud_mean_absolute_error_clear: 1901136.5104 - cloud_psnr: -22.6387 - cloud_root_mean_squared_error: 2770528.4805 - cloud_bandwise_root_mean_squared_error: 1997229.0315 - cloud_mean_absolute_error_covered: 1722210.7075 - cloud_ssim: 0.3470 - cloud_mean_sam_covered: 0.6523 - cloud_mean_sam_clear: 0.5630
   5/6350 [..............................] - ETA: 2:33:35 - loss: 159066224045054459904.0000 - carl_error: 159066224045054459904.0000 - cloud_mean_absolute_error: 79533112022527229952.0000 - cloud_mean_squared_error: inf - cloud_mean_sam: 0.7809 - cloud_mean_absolute_error_clear: 99721074375917830144.0000 - cloud_psnr: -inf - cloud_root_mean_squared_error: inf - cloud_bandwise_root_mean_squared_error: inf - cloud_mean_absolute_error_covered: 59916987194947624960.0000 - cloud_ssim: nan - cloud_mean_sam_covered: 0.7836 - cloud_mean_sam_clear: 0.7331
   6/6350 [..............................] - ETA: 2:17:08 - loss: nan - carl_error: nan - cloud_mean_absolute_error: nan - cloud_mean_squared_error: nan - cloud_mean_sam: 0.6721 - cloud_mean_absolute_error_clear: nan - cloud_psnr: -inf - cloud_root_mean_squared_error: inf - cloud_bandwise_root_mean_squared_error: inf - cloud_mean_absolute_error_covered: nan - cloud_ssim: nan - cloud_mean_sam_covered: 0.6731 - cloud_mean_sam_clear: 0.6335                                                                                                                
   7/6350 [..............................] - ETA: 2:05:18 - loss: nan - carl_error: nan - cloud_mean_absolute_error: nan - cloud_mean_squared_error: nan - cloud_mean_sam: 0.5992 - cloud_mean_absolute_error_clear: nan - cloud_psnr: -inf - cloud_root_mean_squared_error: inf - cloud_bandwise_root_mean_squared_error: inf - cloud_mean_absolute_error_covered: nan - cloud_ssim: nan - cloud_mean_sam_covered: 0.5992 - cloud_mean_sam_clear: 0.5661
   8/6350 [..............................] - ETA: 1:56:30 - loss: nan - carl_error: nan - cloud_mean_absolute_error: nan - cloud_mean_squared_error: nan - cloud_mean_sam: 0.5477 - cloud_mean_absolute_error_clear: nan - cloud_psnr: -inf - cloud_root_mean_squared_error: inf - cloud_bandwise_root_mean_squared_error: inf - cloud_mean_absolute_error_covered: nan - cloud_ssim: nan - cloud_mean_sam_covered: 0.5581 - cloud_mean_sam_clear: 0.4954
   9/6350 [..............................] - ETA: 1:49:39 - loss: nan - carl_error: nan - cloud_mean_absolute_error: nan - cloud_mean_squared_error: nan - cloud_mean_sam: 0.5104 - cloud_mean_absolute_error_clear: nan - cloud_psnr: -inf - cloud_root_mean_squared_error: inf - cloud_bandwise_root_mean_squared_error: inf - cloud_mean_absolute_error_covered: nan - cloud_ssim: nan - cloud_mean_sam_covered: 0.5243 - cloud_mean_sam_clear: 0.4404
  10/6350 [..............................] - ETA: 1:44:07 - loss: nan - carl_error: nan - cloud_mean_absolute_error: nan - cloud_mean_squared_error: nan - cloud_mean_sam: 0.5440 - cloud_mean_absolute_error_clear: nan - cloud_psnr: -inf - cloud_root_mean_squared_error: inf - cloud_bandwise_root_mean_squared_error: inf - cloud_mean_absolute_error_covered: nan - cloud_ssim: nan - cloud_mean_sam_covered: 0.5559 - cloud_mean_sam_clear: 0.4833
  11/6350 [..............................]

**Comment by sllll12** [comment]
> I have retrained this model, and the performance is better now. So I think something wrong in the auther's pretrained file. 我重新训练了模型，结果变好了，至少全白的现象消失了。

Did you modify the model parameters during training? I only changed the batch_size to 1, but the prediction results were poor.
